packages/editor/src/components/provider/index.js

```diff
@@ -24,22 +31,126 @@ import { unlock } from '../../lock-unlock';
+			return;
+		}
+
+		setBlockEditingMode( navigationBlockClientId, 'contentOnly' );
+
+		return () => {
+			unsetBlockEditingMode( navigationBlockClientId );
+		};
+	}, [
+		navigationBlockClientId,
+		unsetBlockEditingMode,
+		setBlockEditingMode,
+	] );
+}
+
+/**
+ * Depending on the post, template and template mode,
+ * returns the appropriate blocks and change handlers for the block editor provider.
+ *
+ * @param {Array}   post         Block list.
+ * @param {boolean} template     Whether the page content has focus (and the surrounding template is inert). If `true` return page content blocks. Default `false`.
+ * @param {boolean} templateMode Whether to wrap the page content blocks in a group block to mimic the post editor. Default `false`.
+ * @return {Array} Block editor props.
+ */
+function useBlockEditorProps( post, template, templateMode ) {
+	const rootLevelPost =
+		!! template && templateMode !== 'hidden' ? template : post;
+	const { type, id } = rootLevelPost;
+	const [ blocks, onInput, onChange ] = useEntityBlockEditor(
+		'postType',
+		type,
+		{ id }
+	);
+	const actualBlocks = useMemo( () => {
+		if ( type === 'wp_navigation' ) {
+			return [
+				createBlock( 'core/navigation', {
+					ref: id,
+					// As the parent editor is locked with `templateLock`, the template locking
+					// must be explicitly "unset" on the block itself to allow the user to modify
+					// the block's content.
+					templateLock: false,
+				} ),
+			];
+		}
+	}, [ type, id ] );
+	const allowRootLevelChanges =
+		!! template && templateMode !== 'disabled' && type === 'wp_navigation';
+	const navigationBlockClientId =
+		type === 'wp_navigation' && actualBlocks && actualBlocks[ 0 ]?.clientId;
+	useForceFocusModeForNavigation( navigationBlockClientId );
+
+	return [
+		actualBlocks ?? blocks,
+		allowRootLevelChanges ? onInput : noop,
+		allowRootLevelChanges ? onChange : noop,
+	];
+}
+
 export const ExperimentalEditorProvider = withRegistryProvider(
 	( {
 		__unstableTemplate,
+		templateMode = 'all',
 		post,
 		settings,
 		recovery,
 		initialEdits,
 		children,
 		BlockEditorProviderComponent = ExperimentalBlockEditorProvider,
 	} ) => {
+		const [ queryContext, setQueryContext ] = useState( { page: 1 } );
+		const rootLevelPost =
+			!! __unstableTemplate && templateMode !== 'hidden'
+				? __unstableTemplate
+				: post;
 		const defaultBlockContext = useMemo( () => {
-			if ( post.type === 'wp_template' ) {
-				return {};
-			}
-			return { postId: post.id, postType: post.type };
-		}, [ post.id, post.type ] );
+			const postContext =
+				post.type !== 'wp_template'
+					? { postId: post.id, postType: post.type }
+					: {};
+
+			return {
+				...postContext,
+				templateSlug:
+					rootLevelPost.type === 'wp_template'
+						? rootLevelPost.slug
+						: undefined,
+				queryContext: [
+					queryContext,
+					( newQueryContext ) =>
+						setQueryContext( {
+							...queryContext,
+							...newQueryContext,
+						} ),
+				],
```

Comment:
The assumption here is that queryContext is just local state, am I wrong here? cc @ntsekouras

Reply:
I think this is obsolete. I can open a PR to remove it, as it requires a couple more changes in block's edit function.

Reply:
PR that removes it here: https://github.com/WordPress/gutenberg/pull/56034